### PR TITLE
Support compress kv

### PR DESCRIPTION
**Pull request by cyita**
## Description

Support Compress KV for mistral and llama.

TODOs:

- [x] llama 4.36-4.41 (except 4.39 &4.40)
- [x] mistral 4.40, 4.41
- [x] mistral 4.37 on mtl
~- [ ] Compress KV with Quant kv~

### 2. User API changes

The compress kv will be activated when 
`export IPEX_LLM_COMPRESS_KV_CACHE=1` and `quantize kv cache is disabled`.

### 4. How to test?
- [ ] N/A
- [x] Unit test: Please manually trigger the PR Validation [here](https://github.com/intel-analytics/ipex-llm-workflow/actions/workflows/llm-PR-validation.yml) by inputting the PR number (e.g., `1234`). And paste your action link here once it has been successfully finished.
- [ ] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/models/mistral.py [hunk 1]
```diff
@@ -1305,3 +1343,68 @@ def mistral_attention_forward_4_39_original(
         attn_weights = None
 
     return attn_output.to(original_dtype), attn_weights, past_key_value
+
+
+def prepare_inputs_for_generation_mistral(
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/mistral.py:1348]: please highlight what are changed for snapkv

File: python/llm/src/ipex_llm/transformers/models/mistral.py [hunk 2]
```diff
@@ -730,7 +737,14 @@ def mistral_attention_forward_4_36_quantized(
...
                     "with a layer index."
                 )
-            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            # [SnapKV] add kv_seq_len
+            if hasattr(self, "kv_seq_len"):
+                if self.kv_seq_len != 0:
+                    kv_seq_len += self.kv_seq_len
+                else:
+                    kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
+            else:
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/mistral.py:746]: Do we still need this?

File: python/llm/src/ipex_llm/transformers/models/mistral.py [hunk 3]
```diff
@@ -1149,11 +1166,15 @@ def mistral_attention_forward_4_39_original(
         kv_seq_len += 1
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/mistral.py:1166]: is `kv_seq_len = cache_k.shape[-2]` above correct?

File: python/llm/src/ipex_llm/transformers/models/mistral.py [hunk 4]
```diff
@@ -1149,11 +1166,15 @@ def mistral_attention_forward_4_39_original(
...
         # update past_key_value's seem_tokens and kv caches.
-        if self.layer_idx == 0:
-            past_key_value._seen_tokens = kv_seq_len
+        # [SnapKV]
+        if use_compresskv:
+            past_key_value.update_seen_tokens(self.layer_idx, q_len)
+            kv_seq_len = past_key_value.get_seq_length()
+        else:
+            if self.layer_idx == 0:
+                past_key_value._seen_tokens = kv_seq_len
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/mistral.py:1175]: Maybe we should always use `DynamicCompressCache` to make the code consistent？

File: python/llm/src/ipex_llm/transformers/models/mistral.py (hunk 3 above)
Comment (cyita) [reply python/llm/src/ipex_llm/transformers/models/mistral.py:1166]: If the compress kv is used, `kv_seq_len` will be set to the correct number at Line1172.

File: python/llm/src/ipex_llm/transformers/models/mistral.py (hunk 2 above)
Comment (cyita) [reply python/llm/src/ipex_llm/transformers/models/mistral.py:746]: The quantized kv version is WIP, I will update it later.

File: python/llm/src/ipex_llm/transformers/models/mistral.py (hunk 4 above)
Comment (cyita) [reply python/llm/src/ipex_llm/transformers/models/mistral.py:1175]: I think it's better to use another cache for example `normal cache` and implement the same interface as the compressed cache to deal with the normal case later.

File: python/llm/src/ipex_llm/transformers/models/mistral.py [hunk 5]
```diff
@@ -15,6 +15,7 @@
 #
 # Some parts of this file is adapted from
 # https://github.com/huggingface/transformers/blob/main/src/transformers/models/mistral/modeling_mistral.py
+# https://github.com/FasterDecoding/SnapKV/blob/main/snapkv/monkeypatch/mistral_hijack_4_37.py
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/models/mistral.py:18]: remove this line?

**Review by rnwang04 — APPROVED** [review]
LGTM

**Comment by cyita** [comment]
PR validation: https://github.com/intel-analytics/ipex-llm-workflow/actions/runs/10091473888

**Pull request merged**